go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,250 @@
+	// ForEachAddressSet calls the given function for each address set
+	// known to the factory
+	ForEachAddressSet(iteratorFn AddressSetIterFunc) error
+}
+
+type ovnAddressSetFactory struct{}
+
+// NewOvnAddressSetFactory creates a new AddressSetFactory backed by
+// address set objects that execute OVN commands
+func NewOvnAddressSetFactory() AddressSetFactory {
+	return &ovnAddressSetFactory{}
+}
+
+// ovnAddressSetFactory implements the AddressSetFactory interface
+var _ AddressSetFactory = &ovnAddressSetFactory{}
+
+// NewAddressSet returns a new address set object
+func (asf *ovnAddressSetFactory) NewAddressSet(name string, ips []net.IP) (AddressSet, error) {
+	return newOvnAddressSet(name, ips)
+}
+
+// ForEachAddressSet will pass the hashedName, namespaceName and
+// the first suffix in the name to the 'iteratorFn' for every address_set in
+// OVN. (Each unhashed name for an ovnAddressSet can be of the form
+// namespaceName.suffix1.suffix2. .suffixN)
+func (asf *ovnAddressSetFactory) ForEachAddressSet(iteratorFn AddressSetIterFunc) error {
+	output, stderr, err := util.RunOVNNbctl("--format=csv", "--data=bare", "--no-heading",
+		"--columns=name,external_ids", "find", "address_set")
+	if err != nil {
+		return fmt.Errorf("error reading address sets: "+
+			"stdout: %q, stderr: %q err: %v", output, stderr, err)
+	}
+	for _, line := range strings.Split(output, "\n") {
+		fmt.Printf("#### %q\n", line)
+		parts := strings.Split(line, ",")
+		if len(parts) != 2 {
+			continue
+		}
+		hashedName := parts[0]
+		for _, externalID := range strings.Fields(parts[1]) {
+			if !strings.HasPrefix(externalID, "name=") {
+				continue
+			}
+			addrSetName := externalID[5:]
+			names := strings.Split(addrSetName, ".")
+			addrSetNamespace := names[0]
+			nameSuffix := ""
+			if len(names) >= 2 {
+				nameSuffix = names[1]
+			}
+			iteratorFn(hashedName, addrSetNamespace, nameSuffix)
+			break
+		}
+	}
+	return nil
+}
+
+// AddressSet is an interface for address set objects
+type AddressSet interface {
+	// GetHashName returns the hashed name of the address set
+	GetHashName() string
+	// GetName returns the descriptive name of the address set
+	GetName() string
+	AddIP(ip net.IP) error
+	DeleteIP(ip net.IP) error
+	Destroy()
+}
+
+type removeFunc func(string)
+
+type ovnAddressSet struct {
+	sync.RWMutex
+	name     string
+	hashName string
+	uuid     string
+	ips      map[string]net.IP
+}
+
+// ovnAddressSet implements the AddressSet interface
+var _ AddressSet = &ovnAddressSet{}
+
+// hash the provided input to make it a valid ovnAddressSet name.
+func hashedAddressSet(s string) string {
+	return hashForOVN(s)
+}
+
+func asDetail(as *ovnAddressSet) string {
+	return fmt.Sprintf("%s/%s/%s", as.uuid, as.name, as.hashName)
+}
+
+func newOvnAddressSet(name string, ips []net.IP) (*ovnAddressSet, error) {
+	as := &ovnAddressSet{
+		name:     name,
+		hashName: hashedAddressSet(name),
+		ips:      make(map[string]net.IP),
```

Comment:
don't know if it's worth changing this to `*net.IP` like I mentioned in my previous comment. `sizeof(reflect.SliceHeader)` > `sizeof(*)`

Reply:
@abhat also means you don't need to copy a whole array of IPs when creating an address set. So sure.